ttnn/cpp/ttnn/operations/embedding/embedding.cpp

```diff
@@ -0,0 +1,81 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttnn/operations/embedding/embedding.hpp"
+#include "ttnn/operations/core/core.hpp"
+#include "ttnn/common/constants.hpp"
+#include "ttnn/operations/embedding/device/embedding_device_operation.hpp"
+#include "ttnn/run_operation.hpp"
+#include "ttnn/operations/data_movement/unsqueeze/unsqueeze.hpp"
+
+namespace ttnn::operations::embedding{
+
+ttnn::Tensor EmbeddingOperation::invoke(
+    uint8_t queue_id,
+    const Tensor& input_tensor_arg,
+    const Tensor& weight_arg,
+    const std::optional<int>& pad_token,
+    const Layout& layout,
+    EmbeddingsType embeddings_type,
+    const std::optional<const DataType> dtype,
+    const std::optional<MemoryConfig>& memory_config,
+    std::optional<Tensor> optional_output_tensor) {
+    if (pad_token.has_value()) {
+        embeddings_type = EmbeddingsType::PADDED;
+    }
+    Tensor mutable_input_tensor = input_tensor_arg;
+    Tensor mutable_weight = weight_arg;
+    if (mutable_input_tensor.get_layout() == ttnn::TILE_LAYOUT) {
+        mutable_input_tensor = ttnn::to_layout(mutable_input_tensor, ttnn::ROW_MAJOR_LAYOUT, std::nullopt, std::nullopt, (Device*)nullptr);
```

Comment:
`(Device*)nullptr` means on host?

Reply:
I thought if I leave it null then it will automatically use the device the previous tensor was on, it's not the case?

Reply:
I am 99% sure it will call on host conversion. It will ignore last parameter anyway.

Reply:
you must call ttnn::untilize/tilize_with_padding if you want to do conversion on device.

Reply:
Ok, so I should retrieve the device from the original tensor and pass that device in.

Reply:
I recommend to call functions  which I mentioned before.